File: examples/matlab_octave/elliptic1DaddScalarBC.m
% Solves the 1D Poisson's equation with Robin boundary conditions
% same example as elliptic1D that uses addScalarBC1D
%

clc
close all

addpath('../../src/matlab_octave')

west = 0;  % Domain's limits
east = 1;

k = 6;  % Operator's order of accuracy
m = 2*k+1;  % Minimum number of cells to attain the desired accuracy
dx = (east-west)/m;  % Step length
grid = [west west+dx/2 : dx : east-dx/2 east]; % grid

% Robin boundary conditions
dc = [1;1];
nc = [1;1];
v = [0;2*exp(1)];

L = lap(k, m, dx);  % 1D Mimetic laplacian operator
U = exp(grid)'; % RHS
[L0,U0] = addScalarBC1D(L,U,k,m,dx,dc,nc,v); % add BC to system
U0 = L0\U0;  % Solve a linear system of equations

% Plot result
plot(grid, U0, 'o')
hold on
plot(grid, exp(grid))
legend('Approximated', 'Analytical', 'Location', 'NorthWest')
title('Poisson''s equation with Robin BC')
xlabel('x')
ylabel('u(x)')

File: julia/MOLE.jl/examples/elliptic/elliptic1DaddScalarBC.jl
"""
A Julia script that solves the 1D Poisson's equation with Robin boundary
conditions using mimetic operators. This example uses addScalarBC

u_xx = e^x

0 < x < 1

Robin Boundary Conditions:
u(0) - u_x(0) = 0
u(1) + u_x(1) = 2 e

Exact Solution:
u(x) = e^x
"""

using LinearAlgebra
using SparseArrays
using Plots

import MOLE: Operators, BCs

# Domain limits
west = 0.0
east = 1.0

k = 6 # operator order of accuracy
m = 2k + 1 # minimum number of cells to attain desired accuracy
dx = (east - west) / m # step length

path = joinpath(@__DIR__, "output") # Output path to store generated plots
mkpath(path)

# 1D staggered grid
grid = [west; (west + dx / 2):dx:(east - dx / 2); east]

# Mimetic Laplacian operator
L = Operators.lap(k, m, dx)

# Robin boundary conditions:  dc*u + nc*(du/dn) = v
dc = (1.0, 1.0)
nc = (1.0, 1.0)
v = (0.0, 2 * exp(1.0))

bc = BCs.ScalarBC1D(dc, nc, v)

# RHS
rhs = exp.(grid)

# Apply boundary conditions and construct new system matrix/array
L0, rhs0 = BCs.addScalarBC!(sparse(L), rhs, bc, k, m, dx)

# Solve linear system
u = L0 \ rhs0

# Plot
p = scatter(grid, u; label = "Approximated", marker = :circle, show = false)
plot!(p, grid, exp.(grid); label = "Analytical", show = false)
plot!(
    p,
    title = "Poisson's equation with Robin BC using addScalarBC",
    xlabel = "x",
    ylabel = "u(x)",
    legend = :topleft,
    show = false,
)
Plots.png(p, joinpath(path, "elliptic1DaddScalarBC_output.png"))
